audiostream2py/data.py

```diff
@@ -0,0 +1,148 @@
+
+    @property
+    def tt(self):
+        """Legacy naming: top time"""
+        return self.end_date
+
+    @staticmethod
+    def concatenate(audio_datas: Sequence['AudioData']):
+        """Join a sequence of AudioData.
+
+        :param audio_datas: list of AudioData
+        :return: The concatenated AudioData.
+        """
+        start_date = audio_datas[0].start_date
+        end_date = audio_datas[-1].end_date
+        waveform = b''.join(ad.waveform for ad in audio_datas)
+        frame_count = sum(ad.frame_count for ad in audio_datas)
+        status_flags = PaStatusFlags.paNoError
+        for ad in audio_datas:
+            status_flags |= PaStatusFlags(ad.status_flags)
+
+        return AudioData(start_date, end_date, waveform, frame_count, status_flags)
+
+    def __lt__(self, other: Union['AudioData', int, float]) -> bool:
+        """Less than comparing start_date
+
+        :param other: AudioData | int | float
+        :return: bool
+        """
+        other_val = other.start_date if isinstance(other, AudioData) else other
+        return self.start_date < other_val
+
+    def __gt__(self, other: Union['AudioData', int, float]) -> bool:
+        """Greater than comparing start_date
+
+        :param other: AudioData | int | float
+        :return: bool
+        """
+        other_val = other.start_date if isinstance(other, AudioData) else other
+        return self.start_date > other_val
+
+    def __getitem__(self, index) -> 'AudioData':
+        """Slice waveform and return AudioData with updated timestamps and data.
+
+        Further info on methodology found in timestamping data stream discussion:
+        https://miro.com/app/board/uXjVPsuJtdM=/?share_link_id=872583858418
+
+        Note: slice 'step' is unlikely used in the traditional sense which is how it is implemented.
+        A future improvement could use 'step' as a channel selector by filtering out bytes of other
+        channels.
+
+        :param index: timestamp or slice
+        :return: Sliced AudioData
+        """
+        if isinstance(index, slice):
+            if index.start is not None:
+                start_sample, start_date = self.nearest_sample_index_and_time(
+                    timestamp=index.start, rounding_type='ceil'
+                )
+            else:
+                start_sample, start_date = 0, self.start_date
+
+            if index.stop is not None:
+                end_sample, end_date = self.nearest_sample_index_and_time(
+                    timestamp=index.stop, rounding_type='floor'
+                )
+            else:
+                end_sample, end_date = self.frame_count, self.end_date
+
+            start = start_sample * self.frame_size
+            stop = end_sample * self.frame_size
+            step = index.step * self.frame_size if index.step is not None else None
+            waveform = self.waveform[start:stop:step]
+            frame_count = end_sample - start_sample
+        else:
```

Comment:
Can be in a separate method too.

Reply:
I made the change but now I'm thinking it's better to completely remove indexing a single sample since there's no good reason to do it like this.